pkg/handler/collector/rekor/rekor.go

```diff
@@ -0,0 +1,331 @@
+	sth := &util.SignedCheckpoint{}
+	first := false
+
+	// Try to read the file containing the previous checkpoints.
+	_, err := os.Stat(logInfoFile)
+
+	// If the file containing previous checkpoints exists.
+	if err == nil {
+		sth, err = readLatestCheckpoint(logInfoFile)
+		if err != nil {
+			return nil, false, fmt.Errorf("reading log info: %v", err)
+		}
+		return sth, first, nil
+	}
+
+	// Otherwise it means that no old snapshot is data available.
+	if errors.Is(err, fs.ErrNotExist) {
+		first = true
+		// Get the latest checkpoint.
+		logInfo, err := mirroring.GetLogInfo(rc.client)
+		if err != nil {
+			return nil, first, fmt.Errorf("getting log info: %v", err)
+		}
+
+		// Parse it.
+		if err := sth.UnmarshalText([]byte(*logInfo.SignedTreeHead)); err != nil {
+			return nil, first, fmt.Errorf("unmarshalling logInfo.SignedTreeHead to Checkpoint: %v", err)
+		}
+
+		return sth, first, nil
+	}
+
+	// If there is any other error while reading the file
+	return nil, false, fmt.Errorf("reading %q: %v", logInfoFile, err)
+
+}
+
+func verifySnapshot(sth *util.SignedCheckpoint, logInfoFile string) error {
+	s, err := sth.SignedNote.MarshalText()
+	if err != nil {
+		return fmt.Errorf("failed to marshal checkpoint: %v", err)
+	}
+
+	// Open file to create new snapshot
+	file, err := os.OpenFile(logInfoFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to open log file: %v", err)
+	}
+	defer file.Close()
+
+	// Replace newlines to flatten checkpoint to single line
+	if _, err := file.WriteString(fmt.Sprintf("%s\n", strings.Replace(string(s), "\n", "\\n", -1))); err != nil {
+		return fmt.Errorf("failed to write to file: %v", err)
+	}
+
+	return nil
+}
+
+func getRekorClient(rekorServerURL string) (*c.Rekor, error) {
+	_, err := url.Parse(rekorServerURL)
+	if err != nil {
+		return nil, err
+	}
+
+	rekorClient, err := client.GetRekorClient(rekorServerURL)
+	if err != nil {
+		return nil, fmt.Errorf("error getting Rekor client: %v", err)
+	}
+
+	return rekorClient, nil
+}
+
+func (rc *rekorCollector) signedCheckpoint2Document(sc *util.SignedCheckpoint) (*processor.Document, error) {
+	// Convert the SignedCheckpoint to bytes.
+	blob, err := marshalSignedCheckpoint(sc)
+	if err != nil {
+		return nil, err
+	}
+
+	// Populate a Document struct.
+	doc := &processor.Document{
+		Blob:   blob,
+		Type:   processor.DocumentITE6SLSA,
+		Format: processor.FormatJSON,
+		SourceInformation: processor.SourceInformation{
+			Collector: string(CollectorRekor),
+			Source:    rc.server,
+		},
+	}
+
+	return doc, nil
+}
+
+func marshalSignedCheckpoint(sc *util.SignedCheckpoint) ([]byte, error) {
+	var scBytes bytes.Buffer
+	encoder := gob.NewEncoder(&scBytes)
+	if err := encoder.Encode(sc); err != nil {
+		return []byte{}, err
+	}
+	return scBytes.Bytes(), nil
+
+}
+
+// ----------RAW COPY PASTE OF REKOR-MONITOR PRIVATE FUNCS----------------------
```

Comment:
can we add a link here for attribution / tracking changes

Reply:
What kind of link would you want to see?

Reply:
the link to the file where the code was copied from will be good.

Reply:
Addresed ✅